Add plugins for server listeners

dmcgowan opened the pull request:
Add new plugin types for server listeners and metrics (which are used by the servers).
- Simplifies server startup
- Relies on plugin lifecycle for all long running processes
- Allows extending containerd's set of listeners and their dependencies
- Separates the server configuration based on the listeners

dmcgowan committed:
Add new server plugin type use for debug server

The new server plugin type allows services which have listeners to be
configured as plugins rather than defined directly in the global
configuration. This provides more configuration consistency and allows
containerd to be extended for new types of api handlers.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Add server plugins for grpc and ttrpc

Migrate configuration and move server initialization

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Add metrics server plugin

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update restart monitor test to not use deprecated address field

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update fuzz to use new config location

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Ensure Plugins is not nil during migration

Signed-off-by: Derek McGowan <derek@mcg.dev>

akerouanton committed:
Default to euid/egid in ttrpc & grpc server plugins

Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>

vvoland commented on cmd/containerd/command/main.go:
```diff
@@ -378,6 +315,33 @@ func applyFlags(cliContext *cli.Context, config *srvconfig.Config) error {
+			config.Plugins["io.containerd.server.v1.ttrpc"] = ttrpcConfig
+		}
+
+		return nil
```
This one will return early and skip `applyPlatformFlags` below, I don't think this was intended?

vvoland commented on cmd/containerd/server/config/config.go:
```diff
@@ -205,42 +212,136 @@ func v1Migrate(ctx context.Context, c *Config) error {
-	Address       string `toml:"address"`
-	GRPCHistogram bool   `toml:"grpc_histogram"`
+	// deprecated: use server plugin io.containerd.server.v1.metrics
+	Address string `toml:"address"`
```
Should this one also be `omitempty` like the others?

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+				serverOpts = append(serverOpts, grpc.MaxSendMsgSize(c.MaxSendMsgSize))
+			}
+
+			// grpcService allows GRPC services to be registered with the underlying server
+			type grpcService interface {
+				Register(*grpc.Server) error
+			}
```
To better decouple the service registration contract from the specific server implementation, it is generally safer to accept the standard `grpc.ServiceRegistrar` interface rather than a concrete `*grpc.Server` pointer. This provides a cleaner boundary, as generated gRPC registration functions already accept `grpc.ServiceRegistrar`.

Note: Transitioning to this interface would require updating the signatures on the implementing plugins.

dmcgowan replied:
Out of scope

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+				serverOpts = append(serverOpts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+			}
+
+			// tcpService allows GRPC services to be registered with the underlying tcp server
+			type tcpService interface {
+				RegisterTCP(*grpc.Server) error
+			}
```
Similarly, it is recommended to use the `grpc.ServiceRegistrar` interface here instead of coupling the components directly to the concrete `*grpc.Server` implementation. This ensures the components rely on the interface contract rather than the underlying structure.

prakleumas commented on cmd/containerd/server/server.go:
```diff
@@ -302,15 +200,12 @@ func New(ctx context.Context, config *srvconfig.Config) (*Server, error) {
+			if !ok {
+				log.G(ctx).WithField("id", id).Warn("plugin does not implement server interface, will not be started")
+			}
+			s.servers = append(s.servers, srv)
```
If the plugin instance does not implement the `server` interface, appending a `nil` interface here introduces the opportunity for a nil pointer dereference when `Start` is later invoked on `s.servers`. Would it be safer to protect the append operation within the type assertion's success branch?

```suggestion
			if srv, ok := instance.(server); ok {
				s.servers = append(s.servers, srv)
			} else {
				log.G(ctx).WithField("id", id).Warn("plugin does not implement server interface, will not be started")
			}
```

prakleumas commented on contrib/fuzz/daemon.go:
```diff
@@ -71,16 +72,10 @@ func startDaemon() {
+			if err := server.Start(ctx); err != nil {
+				log.G(ctx).WithError(err).WithField("address", defaultAddress).Fatal("serve failure")
 			}
 		}()
```
Because `server.Start(ctx)` is now non-blocking (it delegates listener execution to background goroutines via `internal.Serve`), this wrapper goroutine will return almost immediately. This introduces a race condition where `defer server.Stop()` executes instantly, tearing down the daemon before the fuzzer can meaningfully exercise it. Is there a specific reason to run this asynchronously, or would it be safer to bind the shutdown phase to the context's lifecycle?

```suggestion
		go func() {
			defer server.Stop()
			if err := server.Start(ctx); err != nil {
				log.G(ctx).WithError(err).WithField("address", defaultAddress).Fatal("serve failure")
			}
			<-ctx.Done()
		}()
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+					tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+				}
+				serverOpts = append(serverOpts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+			} else if c.TLSCName != "" {
+				tlsConfig, CA, res, err :=
+					wintls.SetupTLSFromWindowsCertStore(ic.Context, c.TLSCName)
+				if err != nil {
+					return nil, fmt.Errorf("failed to setup TLS from Windows cert store: %w", err)
+				}
+				// Cache resource for cleanup (Windows only)
+				setTLSResource(res)
+				if CA != nil {
+					tlsConfig.ClientCAs = CA
+					tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+				}
+				serverOpts = append(serverOpts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+			}
```
This logic bleeds Windows-specific mechanics into a generic payload root. By invoking `wintls.SetupTLSFromWindowsCertStore` here, it causes non-Windows architectures to import the `wintls` package merely to satisfy the `wintls.CertResource` interface contract. It is advisable to sequester this configuration branch entirely behind OS-specific build tags to maintain a strict platform abstraction boundary.

Consider abstracting this operation into `tls_windows.go` and `tls_other.go`.

```suggestion
			} else if c.TLSCName != "" {
				var err error
				serverOpts, err = appendPlatformTLSOpts(ic.Context, serverOpts, c.TLSCName)
				if err != nil {
					return nil, fmt.Errorf("failed to setup platform TLS: %w", err)
				}
			}
```

dmcgowan replied:
Previously done this way, not going to change in this PR

prakleumas commented on plugins/server/grpc/tls_other.go:
```diff
@@ -0,0 +1,28 @@
+
+package grpc
+
+import (
+	"github.com/containerd/containerd/v2/internal/wintls"
+)
+
+// TLS resource helpers are no-ops on other unsupported platforms.
+func setTLSResource(r wintls.CertResource) {}
+
+func cleanupTLSResources() {}
```
Importing a Windows-specific internal package (`wintls`) in a `!windows` build file compromises the platform isolation contract. Non-Windows architectures ideally do not depend on Windows types for compilation. If the Windows TLS initialization is fully encapsulated within `tls_windows.go`, this file will not need to reference `wintls.CertResource`.

```suggestion
// Platform specific TLS initialization and cleanup can be handled here without importing Windows packages.
func appendPlatformTLSOpts(ctx context.Context, opts []grpc.ServerOption, tlsCName string) ([]grpc.ServerOption, error) {
	// No-op for non-Windows
	return opts, nil
}

func cleanupTLSResources() {}
```

prakleumas commented on cmd/containerd/server/config/config.go:
```diff
@@ -205,42 +212,136 @@ func v1Migrate(ctx context.Context, c *Config) error {
+		c.Debug.UID = 0
+		c.Debug.GID = 0
+	}
+	grpcAddress := c.GRPC.Address
```
The legacy configuration logic explicitly falls back to using the GRPC UID and GID for the TTRPC server if TTRPC is not configured. Because `c.GRPC.UID` and `c.GRPC.GID` are cleared during the GRPC migration below, they are lost by the time the TTRPC migration occurs. It is necessary to capture them here so they can be inherited in the TTRPC configuration.

```suggestion
	grpcAddress := c.GRPC.Address
	grpcUID := c.GRPC.UID
	grpcGID := c.GRPC.GID
```

prakleumas commented on cmd/containerd/server/config/config.go:
```diff
@@ -205,42 +212,136 @@ func v1Migrate(ctx context.Context, c *Config) error {
+			"max_send_message_size": c.GRPC.MaxSendMsgSize,
+		}
+		c.GRPC.TCPAddress = ""
+		c.GRPC.TCPTLSCA = ""
+		c.GRPC.TCPTLSCert = ""
+		c.GRPC.TCPTLSKey = ""
+		c.GRPC.TCPTLSCName = ""
+		c.GRPC.MaxRecvMsgSize = 0
+		c.GRPC.MaxSendMsgSize = 0
+	}
+	if c.Plugins["io.containerd.server.v1.ttrpc"] == nil {
+		ttrpcAddress := c.TTRPC.Address
+		if ttrpcAddress == "" && grpcAddress != "" {
+			ttrpcAddress = grpcAddress + ".ttrpc"
+		}
+		if ttrpcAddress != "" || c.TTRPC.UID != 0 || c.TTRPC.GID != 0 {
+			c.Plugins["io.containerd.server.v1.ttrpc"] = map[string]any{
+				"address": ttrpcAddress,
+				"uid":     c.TTRPC.UID,
+				"gid":     c.TTRPC.GID,
+			}
+			c.TTRPC.Address = ""
+			c.TTRPC.UID = 0
+			c.TTRPC.GID = 0
+		}
+	}
```
Apply the captured GRPC UID and GID fallbacks to maintain parity with the legacy initialization contract, ensuring the TTRPC socket inherits the correct ownership when its configuration is omitted.

```suggestion
		ttrpcAddress := c.TTRPC.Address
		ttrpcUID := c.TTRPC.UID
		ttrpcGID := c.TTRPC.GID
		if ttrpcAddress == "" && grpcAddress != "" {
			ttrpcAddress = grpcAddress + ".ttrpc"
			ttrpcUID = grpcUID
			ttrpcGID = grpcGID
		}
		if ttrpcAddress != "" || ttrpcUID != 0 || ttrpcGID != 0 {
			c.Plugins["io.containerd.server.v1.ttrpc"] = map[string]any{
				"address": ttrpcAddress,
				"uid":     ttrpcUID,
				"gid":     ttrpcGID,
			}
			c.TTRPC.Address = ""
			c.TTRPC.UID = 0
			c.TTRPC.GID = 0
		}
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+
+			if p, err := ic.GetByID(plugins.MetricsPlugin, "grpc-prometheus"); err == nil {
+				prometheusServerMetrics = p.(*grpc_prometheus.ServerMetrics)
+				streamOpts = append(streamOpts, prometheusServerMetrics.StreamServerInterceptor())
+				unaryOpts = append(unaryOpts, prometheusServerMetrics.UnaryServerInterceptor())
+				prometheusServerMetrics.InitializeMetrics(nil)
+			}
```
It looks like `prometheusServerMetrics.InitializeMetrics(nil)` is called here, and then correctly called again with the instantiated server on line 238. Passing `nil` to `InitializeMetrics` introduces the opportunity for a panic if the library attempts to inspect the server's registered services. Should this early call be removed?

```suggestion
				streamOpts = append(streamOpts, prometheusServerMetrics.StreamServerInterceptor())
				unaryOpts = append(unaryOpts, prometheusServerMetrics.UnaryServerInterceptor())
			}
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+func (s grpcServer) Start(ctx context.Context) error {
+	log.G(ctx).Infof("starting GRPC server on %s with %d/%d", s.config.Address, s.config.UID, s.config.GID)
```
We generally prefer structured logs rather than printf style formatting for dynamic variables. It makes logs easier to query and index.

```suggestion
	log.G(ctx).WithField("address", s.config.Address).WithField("uid", s.config.UID).WithField("gid", s.config.GID).Info("starting GRPC server")
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+			ps, err := ic.GetByType(plugins.GRPCPlugin) // ensure grpc plugin is initialized
+			if err != nil {
+				return nil, err
+			}
```
It is generally safer to gracefully handle the case where no plugins are found, ensuring we don't inadvertently fail server initialization if `plugin.ErrPluginNotFound` is returned. This matches the resilient handling applied in the `ttrpc` server setup.

```suggestion
			if err != nil && !errors.Is(err, plugin.ErrPluginNotFound) {
				return nil, err
			}
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+			ps, err := ic.GetByType(plugins.GRPCPlugin) // ensure grpc plugin is initialized
+			if err != nil {
+				return nil, err
+			}
```
Similar to the main grpc initialization, it is generally safer to tolerate an absence of plugins to prevent hard failures during startup if `plugin.ErrPluginNotFound` is returned.

```suggestion
			if err != nil && !errors.Is(err, plugin.ErrPluginNotFound) {
				return nil, err
			}
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+				return nil, err
+			}
+			for _, p := range ps {
+				if gs, isGRPC := p.(grpcService); isGRPC {
```
nit: use `ok` for type assertions

```suggestion
				if gs, ok := p.(grpcService); ok {
```

prakleumas commented on plugins/server/grpc/plugin.go:
```diff
@@ -0,0 +1,293 @@
+			}
+			var hasService bool
+			for _, p := range ps {
+				if gs, isGRPC := p.(tcpService); isGRPC {
```
nit: use `ok` for type assertions

```suggestion
				if gs, ok := p.(tcpService); ok {
```

prakleumas commented on plugins/server/ttrpc/plugin.go:
```diff
@@ -0,0 +1,135 @@
+				return nil, err
+			}
+			for _, p := range ps {
+				if gs, isGRPC := p.(ttrpcService); isGRPC {
```
nit: use `ok` for type assertions

```suggestion
				if gs, ok := p.(ttrpcService); ok {
```

prakleumas commented on plugins/server/ttrpc/plugin.go:
```diff
@@ -0,0 +1,135 @@
+				return nil, err
+			}
+			for _, p := range ps {
+				if gs, isGRPC := p.(ttrpcService); isGRPC {
```
nit: use `ok` for type assertions

```suggestion
				if gs, ok := p.(ttrpcService); ok {
```

prakleumas commented on cmd/containerd/command/main.go:
```diff
@@ -378,6 +315,33 @@ func applyFlags(cliContext *cli.Context, config *srvconfig.Config) error {
+
+		_, ok = config.Plugins["io.containerd.server.v1.ttrpc"]
+		if !ok {
+			ttrpcConfig = map[string]any{
+				"address": s + ".ttrpc",
+				"uid":     grpcConfig["uid"],
+				"gid":     grpcConfig["gid"],
+			}
```
If `uid` or `gid` were not explicitly set in the user's configuration, `grpcConfig["uid"]` and `grpcConfig["gid"]` will evaluate to `nil` here (as plugin defaults have not yet been applied). Injecting explicit `nil` values into the `ttrpcConfig` map might cause unexpected behavior when the plugin configuration is later decoded into its struct. Would it be safer to conditionally copy these fields only if they are present?

```suggestion
			ttrpcConfig = map[string]any{
				"address": s + ".ttrpc",
			}
			if uid, ok := grpcConfig["uid"]; ok {
				ttrpcConfig["uid"] = uid
			}
			if gid, ok := grpcConfig["gid"]; ok {
				ttrpcConfig["gid"] = gid
			}
```

prakleumas commented on plugins/server/metrics/plugin.go:
```diff
@@ -0,0 +1,80 @@
+			m.Handle("/v1/metrics", metrics.Handler())
+
+			return server{
+				handler: m,
+				address: c.Address,
+			}, nil
+		},
```
Returning the struct by value prevents methods from securely holding state modifications, such as storing the initialized `http.Server` instance. It is generally safer to return a pointer so that you can retain server bounds and implement an `io.Closer` that shuts down the active service cleanly.

```suggestion
			return &server{
				handler: m,
				address: c.Address,
			}, nil
```

prakleumas commented on plugins/server/metrics/plugin.go:
```diff
@@ -0,0 +1,80 @@
+			}, nil
+		},
+	})
+}
+
+type server struct {
+	handler http.Handler
+	address string
+}
+
+func (s server) Start(ctx context.Context) error {
+	l, err := net.Listen("tcp", s.address)
+	if err != nil {
+		return fmt.Errorf("failed to get listener for metrics endpoint: %w", err)
+	}
+	srv := &http.Server{
+		Handler:           s.handler,
+		ReadHeaderTimeout: 5 * time.Minute, // "G112: Potential Slowloris Attack (gosec)"; not a real concern for our use, so setting a long timeout.
+	}
+	internal.Serve(ctx, l, srv.Serve)
+
```
Because `Start` is now non-blocking and relies on `internal.Serve` (which spawns a background goroutine), this localized initialization introduces the opportunity for the underlying TCP listener to remain open and the goroutine to leak upon a graceful daemon shutdown. It is generally safer to maintain a reference to the `http.Server` and implement `io.Closer`.

```suggestion
type server struct {
	handler http.Handler
	address string
	srv     *http.Server
}

func (s *server) Start(ctx context.Context) error {
	l, err := net.Listen("tcp", s.address)
	if err != nil {
		return fmt.Errorf("failed to get listener for metrics endpoint: %w", err)
	}
	s.srv = &http.Server{
		Handler:           s.handler,
		ReadHeaderTimeout: 5 * time.Minute, // "G112: Potential Slowloris Attack (gosec)"; not a real concern for our use, so setting a long timeout.
	}
	internal.Serve(ctx, l, s.srv.Serve)

	return nil
}

func (s *server) Close() error {
	if s.srv != nil {
		return s.srv.Close()
	}
	return nil
}
```

prakleumas commented on plugins/server/debug/plugin.go:
```diff
@@ -0,0 +1,98 @@
+	internal.Serve(ctx, l, s.handler.Serve)
+
+	return nil
+}
```
Because `internal.Serve` initiates an asynchronous listener, omitting an `io.Closer` implementation introduces the opportunity for a socket and goroutine leak during daemon shutdown. It is generally safer to implement `Close() error` to ensure the HTTP server is terminated cleanly.

```suggestion
}

func (s server) Close() error {
	if s.handler != nil {
		return s.handler.Close()
	}
	return nil
}
```

prakleumas commented on cmd/containerd/server/server.go:
```diff
@@ -575,9 +416,22 @@ func (pc *proxyClients) getClient(address string) (*grpc.ClientConn, error) {
-	if err == nil || errors.Is(err, net.ErrClosed) {
-		return nil
+func readString(properties map[string]any, key ...string) string {
+	for _, k := range key {
+		if v, ok := properties[k]; ok {
+			switch v := v.(type) {
+			case map[string]any:
+				properties = v
+				continue
+			case string:
+				return v
+			case fmt.Stringer:
+				return v.String()
+			case []byte:
+				return string(v)
+			}
+		}
```
It appears that `readString` will eagerly return a value if an intermediate key in the path resolves to a `string`, `fmt.Stringer`, or `[]byte`. For example, if a configuration mistakenly provides a string for a parent key, this function returns that string instead of continuing the search or returning an empty string. Could we ensure that it only returns a leaf value if it has successfully reached the final key in the path?

```suggestion
	for i, k := range key {
		if v, ok := properties[k]; ok {
			switch v := v.(type) {
			case map[string]any:
				properties = v
				continue
			case string:
				if i == len(key)-1 {
					return v
				}
			case fmt.Stringer:
				if i == len(key)-1 {
					return v.String()
				}
			case []byte:
				if i == len(key)-1 {
					return string(v)
				}
			}
		}
```

prakleumas commented on cmd/containerd/server/config/config.go:
```diff
@@ -205,42 +212,136 @@ func v1Migrate(ctx context.Context, c *Config) error {
 	return nil
 }
 
+func serviceMigrate(ctx context.Context, c *Config) error {
```
Can we add a new test that covers the correctness of `serviceMigrate`? This is implicitly covered by higher-level configuration test flows, but it'd be good to validate at least some of the basic property mapping logic (e.g. ensuring `GRPC` and `TTRPC` settings migrate accurately to the new plugins map) directly in a unit test.

prakleumas commented on cmd/containerd/server/config/config.go:
```diff
@@ -94,6 +89,18 @@ type Config struct {
+	TTRPC TTRPCConfig `toml:"ttrpc,omitempty"`
+	// Metrics and monitoring settings
+	// deprecated: use server plugin io.containerd.server.v1.metrics
+	Metrics MetricsConfig `toml:"metrics,omitempty"`
```
To establish a formal contract with standard Go static analysis tools (such as `go doc` and `go vet`), deprecation notices rely on the exact prefix `Deprecated: ` (with a capital 'D'). It is generally safer to align with this convention to ensure IDEs and downstream consumers accurately recognize these fields as phased out during the version 3 to 4 lifecycle transition.

```suggestion
	// Deprecated: use server plugins io.containerd.server.v1.grpc and io.containerd.server.v1.grpc-tcp
	GRPC GRPCConfig `toml:"grpc,omitempty"`
	// TTRPC configuration settings
	// Deprecated: use server plugin io.containerd.server.v1.ttrpc
	TTRPC TTRPCConfig `toml:"ttrpc,omitempty"`
	// Metrics and monitoring settings
	// Deprecated: use server plugin io.containerd.server.v1.metrics
	Metrics MetricsConfig `toml:"metrics,omitempty"`
```

(It is recommended to apply this capitalization to the field-level comments inside the `GRPCConfig`, `TTRPCConfig`, `Debug`, and `MetricsConfig` structs as well.)

prakleumas commented on plugins/server/metrics/plugin.go:
```diff
@@ -0,0 +1,80 @@
+	registry.Register(&plugin.Registration{
+		Type: plugins.ServerPlugin,
+		ID:   "metrics",
+		Requires: []plugin.Type{
+			plugins.HTTPHandler,
+		},
```
Is there a specific reason `plugins.HTTPHandler` is listed as a requirement here? Unlike the debug server plugin, I don't see this plugin querying the initialization context for HTTP handlers. If it's not strictly necessary, it is generally safer to omit the requirement to prevent unintentional initialization ordering constraints.

```suggestion
```

prakleumas commented on cmd/containerd/server/server.go:
```diff
@@ -148,104 +136,14 @@ func New(ctx context.Context, config *srvconfig.Config) (*Server, error) {
-		required    = make(map[string]struct{})
+		initialized  = plugin.NewPluginSet()
+		required     = make(map[string]struct{})
+		grpcAddress  = readString(config.Plugins, "io.containerd.server.v1.grpc", "address")
+		ttrpcAddress = readString(config.Plugins, "io.containerd.server.v1.ttrpc", "address")
```
What is the expected behavior for `ttrpcAddress` if a user provides a v4 configuration that specifies a custom GRPC address but omits the TTRPC plugin block entirely? Previously, the TTRPC address dynamically tracked the GRPC address (`grpcAddress + ".ttrpc"`). Without that fallback here, `readString` will return an empty string, meaning `plugins.PropertyTTRPCAddress` will be empty for dependent plugins. Additionally, the TTRPC server plugin itself may bind to the global default rather than tracking the custom GRPC address. It might be worth re-introducing a fallback mechanism to guarantee they stay linked when unconfigured.

dmcgowan replied:
I think this is a reasonable change for v4 configuration. Since any migrated configuration would keep the same behavior, this would not be a breaking change, since v4 is a new configuration version. I'll make sure to document this, that the values are now independent and when ttrpc is not set, it will default to the ttrpc default, not derived from the grpc address.

prakleumas reviewed (commented):
<sup>beep boop</sup>
Hello! I am @samuelkarp's friendly code review robot 🤖! Sam asked me to take a look and post my comments directly on GitHub. 

Please treat my comments as suggestions and advisory, not mandates. Since I am an LLM and my output is used here directly, you should evaluate carefully and see if my suggestions make sense. Feel free to directly resolve any you disagree with.

<sup>This is _not_ an automated action. This bot does not read or reply to comments.</sup>

dmcgowan committed:
Add tests for serviceMigrate config migration

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update docs for version 4 server plugin configuration

Document the new server plugin configuration blocks for GRPC, TTRPC,
debug, and metrics. Mark the legacy top-level sections as deprecated.
Note that in version 4, the TTRPC plugin is configured independently
from GRPC and uses its own defaults when its plugin block is omitted.

Signed-off-by: Derek McGowan <derek@mcg.dev>

austinvazquez commented on cmd/containerd/server/config/config.go:
```diff
@@ -205,42 +216,155 @@ func v1Migrate(ctx context.Context, c *Config) error {
+			"tls_key":               c.GRPC.TCPTLSKey,
+			"tls_common_name":       c.GRPC.TCPTLSCName,
+			"max_recv_message_size": c.GRPC.MaxRecvMsgSize,
+			"max_send_message_size": c.GRPC.MaxSendMsgSize,
```
If address and tcp address are both set, then the user config for max msg size will be dropped here.

fuweid commented on plugins/server/internal/serve.go:
```diff
@@ -0,0 +1,39 @@
+	go func() {
+		defer l.Close()
+
+		if err := serveFunc(l); err != nil && !errors.Is(err, net.ErrClosed) && !errors.Is(err, http.ErrServerClosed) && !errors.Is(err, ttrpc.ErrServerClosed) {
```
maybe we need to helper function for that.

fuweid replied:
We can handle it in the follow-up

fuweid reviewed (approved):
LGTM

Nice refactor.

---

Follow-up:

It looks like it migrate before merge

there is case

```
# root.toml

version = 3
root = "/tmp/pr12562-repro-grpc/root"
state = "/tmp/pr12562-repro-grpc/state"
imports = ["dropin.toml"]

[grpc]
  address = "/tmp/pr12562-grpc.sock"
  uid = 0
  gid = 0

# dropin.toml
version = 3

[grpc]
  max_send_message_size = 33554432
```

It will show

```
WARN[2026-04-22T13:26:57.901574474-04:00] failed to load plugin                         error="grpc address cannot be empty: invalid argument" id=io.containerd.server.v1.grpc type=io.containerd.server.v1
```

akhilerm commented:
> LGTM
> 
> Nice refactor.
> 
> Follow-up:
> 
> It looks like it migrate before merge
> 
> there is case
> 
> ```
> # root.toml
> 
> version = 3
> root = "/tmp/pr12562-repro-grpc/root"
> state = "/tmp/pr12562-repro-grpc/state"
> imports = ["dropin.toml"]
> 
> [grpc]
>   address = "/tmp/pr12562-grpc.sock"
>   uid = 0
>   gid = 0
> 
> # dropin.toml
> version = 3
> 
> [grpc]
>   max_send_message_size = 33554432
> ```
> 
> It will show
> 
> ```
> WARN[2026-04-22T13:26:57.901574474-04:00] failed to load plugin                         error="grpc address cannot be empty: invalid argument" id=io.containerd.server.v1.grpc type=io.containerd.server.v1
> ```

@fuweid This caused https://github.com/containerd/containerd/issues/13316. When any non empty value in `[grpc]`  is set in older configs and address is not set, address goes to empty.

AkihiroSuda commented:
Regression:
- https://github.com/containerd/containerd/issues/13362
